### PR TITLE
chore(hcl2cdk): prepare for convert using Terraform function bindings

### PR DESCRIPTION
**Pull request by ansgarm**
- feat: generate functions map to be used by convert
- feat: add overrides for functions and type information
- feat: add functions.test.ts containing test cases and example implementation for how the functions map is going to be used
- feat: value is available in meta now
- feat: add range and lookup function overrides

(extracted from #2657 to have something small and incremental that we can merge)

### CODE REVIEW

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

File: packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+      throw new Error("Unsupported type: " + (tfAst as any).type);
+  }
+}
+
+function terraformScopeTraversalToTs(
+  tfAst: ScopeTraversal,
+  _targetType: AttributeType | undefined
+): t.Expression {
+  return t.identifier(
+    "TodoReference-" + tfAst.meta.traversal.map((t) => t.segment).join("-")
```
Comment (DanielMSchmidt) [inline packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts:636]: Seems like a todo is left here

File: packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+  throw new Error("Template currently only supports exactly one child");
+}
+
+function terraformLiteralValueToTs(
+  tfAst: LiteralValue,
+  targetType: AttributeType | undefined
+): t.Expression {
+  const literalExpression = t.stringLiteral(tfAst.meta.value);
+  return coerceType(
+    // TODO: scope should be passed instead, even if coerceType won't need it probably
```
Comment (DanielMSchmidt) [inline packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts:656]: TODO left

File: packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+      }
+    }
+  });
+
+  const returnType = mapping.returnType;
+
+  const callExpression = t.callExpression(callee, args);
+
+  return coerceType(
+    // TODO: scope should be passed instead, even if coerceType won't need it probably
```
Comment (DanielMSchmidt) [inline packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts:724]: TODO left

File: packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts (hunk 1 above)
Comment (ansgarm) [reply packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts:636]: That's on purpose, as this work is not yet feature complete (that's also why all of that code is currently still living in the testcase). I intend the next PR to move that to its own file and e.g. follow up on the reference parsing bits and other open todos.

File: packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts (hunk 2 above)
Comment (ansgarm) [reply packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts:656]: This is for later, when this code is pulled out into it's own file – this PR is only a preparation for that work.

File: packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts (hunk 3 above)
Comment (ansgarm) [reply packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts:724]: same reasoning as above

**Review by DanielMSchmidt — APPROVED** [review]

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Pull request merged**